[PyTorch] Fix waste in unfold()

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55339 [PyTorch] Fix waste in unfold()**
* #55336 [PyTorch] Cache self.size(dim) in TensorShape functions

Use DimVector. Avoid calling size()/stride() when we know argument is in bounds.

Differential Revision: [D27577647](https://our.internmc.facebook.com/intern/diff/D27577647/)

swolchok committed:
[PyTorch] Fix waste in unfold()

Use DimVector. Avoid calling size()/stride() when we know argument is in bounds.

Differential Revision: [D27577647](https://our.internmc.facebook.com/intern/diff/D27577647/)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2133,19 +2133,22 @@ Tensor alias(const Tensor& self) {
+
+  const auto sizes = self.sizes();
+  const auto strides = self.strides();
+  int64_t max_size = self.dim() == 0 ? 1 : sizes[0];
```
it used to be `sizes[dimension]`, why the change?

ngimel replied:
Also, `fold` accepts negative dimension, so you have to do `maybe_wrap_dim`

swolchok replied:
it's wrapped on line 2134 above (the first line of the function), isn't it?

ngimel replied:
Oh, my bad, I'm blind.

swolchok committed:
fix accidental change from dimension to 0 on "[PyTorch] Fix waste in unfold()"


Use DimVector. Avoid calling size()/stride() when we know argument is in bounds.

Differential Revision: [D27577647](https://our.internmc.facebook.com/intern/diff/D27577647/)

[ghstack-poisoned]
